### PR TITLE
Bug 925424 - Only sync the versions we need r=wlach

### PR DESCRIPTION
**Pull request by lambdabaa**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,11 +12,11 @@ npm install -g npm-mirror
...
-DEBUG=* bin/npm-mirror \
-  --host http://secret-npm-mirror.com \
-  --manifestFile /path/to/target/package.json \
-  --registry http://registry.npmjs.org \
-  --packageDir /where/we/put/packages/
+DEBUG=npm-mirror:SyncManager bin/npm-mirror \
+  --master http://registry.npmjs.org \
+  --manifest /path/to/target/package.json \
+  --hostname http://secret-npm-mirror.com \
+  --root /where/we/put/packages/
```
Comment (lightsofapollo) [inline README.md:24]: ++ these new arg names are much better

File: lib/download.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/*global cache*/
+var debug = require('debug')('npm-mirror:Download'),
+    fs = require('graceful-fs'),
+    http = require('http');
+
+
+if (!global.cache) {
```
Comment (lightsofapollo) [inline lib/download.js]: ? my immediate response is this seems like a bad idea can you add a comment why this is a global not as tracked as state in some object?

File: lib/download.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return callback && callback(
+          new Error('Bad status ' + res.statusCode + ' for ' + url));
+      }
+
+      res.setEncoding('utf-8');
+      var result = '';
+      res.on('data', function(data) {
+        result += data;
+      });
+      res.on('end', function() {
```
Comment (lightsofapollo) [inline lib/download.js:60]: hmm... do we handle circular dependencies at all?

a => b => c
b => d => c

File: lib/syncmanager.js [hunk 4]
```diff
@@ -1,354 +1,370 @@
...
 
-        var packageRoot = JSON.parse(packageRootData);
-        var versions = Object.keys(packageRoot.versions);
-        this.syncPackageVersions(packageName, versions, function() {
-          this.savePackageRoot(packageName, versions, callback);
+        // Careful! The version could still be a git/tarball url...
+        // Or version could be null if semver couldn't find a match.
+        if (!version || version === 'null' || UrlCheck.isWebUrl(version) ||
+            UrlCheck.isGitUrl(version)) {
+          // TODO(gaye): Eh? What to do?
```
Comment (lightsofapollo) [inline lib/syncmanager.js:126]: look at how npm handles this internally... its entirely possible we are redoing what npm has already done for us and we can simply reuse that logic.

File: lib/syncmanager.js [hunk 5]
```diff
@@ -1,354 +1,370 @@
...
+          return process();
+        }
+
+        var packageVersionUrl = Package.url(this.master, package, version);
+        Download.download(packageVersionUrl, function(e, packageVersionData) {
+          if (e) {
+            // If we choke on a package, it's probably something wacky or
+            // no longer exists. Let's ignore it for now, but log it just
+            // in case.
+            debug(e);
```
Comment (lightsofapollo) [inline lib/syncmanager.js]: you might want to prefix e with something:

``` js
debug('error downloading package', pkgName, e);
```

File: lib/urlcheck.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+
+var UrlCheck = {
+  /**
+   * @param {string} str string to check.
+   * @return {boolean} whether or not it's a git url.
+   */
+  isGitUrl: function(str) {
+    // TODO(gaye): This is a terrible implementation.
+    if (S(str).startsWith('git://')) {
```
Comment (lightsofapollo) [inline lib/urlcheck.js]: =p startsWith will exist someday until then indexOf should be fine?

``` js
if (str.indexOf('git://') === 0) {
}
```

File: package.json [hunk 7]
```diff
@@ -12,6 +12,8 @@
     "commander": "~2.0.0",
     "debug": "~0.7.2",
     "graceful-fs": "~2.0.1",
+    "semver": "~2.1.0",
+    "string": "~1.6.0",
```
Comment (lightsofapollo) [inline package.json]: not sure you need this

File: test/integration/sync_test.js [hunk 8]
```diff
@@ -1,86 +1,9 @@
-/*global __dirname, path, fs, require, test, suite, suiteSetup, suiteTeardown,
```
Comment (lightsofapollo) [inline test/integration/sync_test.js:1]: I hope we are going to add more tests here?

Comment (lambdabaa) [inline test/integration/sync_test.js:1]: Yes definitely. I always write tests last... :(

File: lib/download.js (hunk 3 above)
Comment (lambdabaa) [inline lib/download.js:60]: As discussed on IRC, when we find "new" dependencies in our graph search, we add them to the graph. Otherwise we ignore them.

Comment (lambdabaa) [inline lib/download.js:60]: The graph state gets held SyncManager#packageToVersions btw

File: lib/syncmanager.js (hunk 4 above)
Comment (lambdabaa) [inline lib/syncmanager.js:126]: I'll go take a look

File: lib/syncmanager.js (hunk 5 above)
Comment (lambdabaa) [inline lib/syncmanager.js]: ++

File: lib/urlcheck.js (hunk 6 above)
Comment (lambdabaa) [inline lib/urlcheck.js]: Haha ok fine

File: package.json (hunk 7 above)
Comment (lambdabaa) [inline package.json]: I felt weird about it too...

File: test/integration/sync_test.js (hunk 8 above)
Comment (lambdabaa) [inline test/integration/sync_test.js:1]: More tests are here! (in integration/sync_test.js)

File: bin/npm-mirror [hunk 9]
```diff
@@ -25,59 +26,61 @@ function parseManifest(manifestFile) {
...
+  var hostname = program.hostname;
+  check(hostname).isUrl();
+  var manifest = parseManifest(program.manifest);
+  // We shouldn't sync the manifest's package since we want
+  // the mirror to be able to sync the dependencies of private packages.
+  var packageToVersions = Package.dependencies(manifest);
+  var master = program.master;
+  check(master).isUrl();
+  var root = resolveRoot(program.root);
+  console.log(master);
```
Comment (lightsofapollo) [inline bin/npm-mirror]: lots of debug logging?

File: lib/download.js (hunk 2 above)
Comment (lightsofapollo) [inline lib/download.js]: ^^^ this is automatic r- from me we should not be using globals

File: bin/npm-mirror (hunk 9 above)
Comment (lambdabaa) [inline bin/npm-mirror]: Ah I meant to delete this!

File: lib/download.js (hunk 2 above)
Comment (lambdabaa) [inline lib/download.js]: global -> singleton :)

**Pull request merged**